File: orchestration-os-merged/scripts/python/agent.py
#!/usr/bin/env python3
"""
Python AI Agent - CLI Tool
Accepts natural language prompts and executes Python code
"""

import os
import sys
import json
import subprocess
import re
from typing import Optional

OLLAMA_HOST = os.environ.get("OLLAMA_HOST", "http://localhost:11434")
OLLAMA_MODEL = os.environ.get("OLLAMA_MODEL", "qwen2.5-coder:3b")
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY", "")

def call_ollama(prompt: str) -> Optional[str]:
    """Call Ollama for code generation"""
    import urllib.request
    import urllib.error
    
    payload = {
        "model": OLLAMA_MODEL,
        "messages": [{"role": "user", "content": prompt}],
        "stream": False
    }
    
    try:
        req = urllib.request.Request(
            f"{OLLAMA_HOST}/api/chat",
            data=json.dumps(payload).encode(),
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        with urllib.request.urlopen(req, timeout=120) as response:
            result = json.loads(response.read().decode())
            return result.get("message", {}).get("content", "")
    except Exception as e:
        print(f"Ollama error: {e}", file=sys.stderr)
        return None

def call_gemini(prompt: str) -> Optional[str]:
    """Call Gemini API for code generation"""
    if not GEMINI_API_KEY:
        return None
    
    import urllib.request
    import urllib.error
    
    payload = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {"temperature": 0.7, "maxOutputTokens": 8192}
    }
    
    try:
        req = urllib.request.Request(
            f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key={GEMINI_API_KEY}",
            data=json.dumps(payload).encode(),
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        with urllib.request.urlopen(req, timeout=60) as response:
            result = json.loads(response.read().decode())
            return result.get("candidates", [{}])[0].get("content", {}).get("parts", [{}])[0].get("text", "")
    except Exception as e:
        print(f"Gemini error: {e}", file=sys.stderr)
        return None

def generate_code(prompt: str) -> str:
    """Generate Python code from prompt using available LLM"""
    full_prompt = f"""You are an expert Python programmer. Generate ONLY Python code for this request:

{prompt}

RULES:
- Output ONLY Python code, nothing else
- NO explanations, NO comments describing what the code does
- NO markdown formatting, NO backticks, NO code block markers
- Code must be valid Python syntax
- Code must be complete and runnable

Output just the raw Python code now:"""

    # Try Ollama first
    result = call_ollama(full_prompt)
    if result:
        return clean_code(result)
    
    # Fallback to Gemini
    result = call_gemini(full_prompt)
    if result:
        return clean_code(result)
    
    raise Exception("No LLM available - configure OLLAMA_HOST or GEMINI_API_KEY")

def clean_code(raw: str) -> str:
    """Clean markdown formatting and extract code"""
    # Remove triple backticks and language specifiers
    code = re.sub(r'^```python\s*', '', raw, flags=re.MULTILINE)
    code = re.sub(r'^```\s*$', '', code, flags=re.MULTILINE)
    code = re.sub(r'^```py\s*', '', code, flags=re.MULTILINE)
    # Remove leading backticks
    code = re.sub(r'^`+', '', code, flags=re.MULTILINE)
    # Remove trailing backticks
    code = re.sub(r'`+$', '', code, flags=re.MULTILINE)
    return code.strip()

def execute_code(code: str) -> tuple[str, str, int]:
    """Execute Python code and return stdout, stderr, returncode"""
    try:
        result = subprocess.run(
            ["python3", "-c", code],
            capture_output=True,
            text=True,
            timeout=30
        )
        return result.stdout, result.stderr, result.returncode
    except subprocess.TimeoutExpired:
        return "", "Execution timed out (30s limit)", 124
    except Exception as e:
        return "", str(e), 1

def main():
    if len(sys.argv) < 2:
        print("Usage: python agent.py <prompt>", file=sys.stderr)
        print("Example: python agent.py 'Print hello world 10 times'", file=sys.stderr)
        sys.exit(1)
    
    prompt = " ".join(sys.argv[1:])
    print(f"[Agent] Generating code for: {prompt}", file=sys.stderr)
    
    try:
        # Generate code
        code = generate_code(prompt)
        print(f"[Agent] Generated {len(code)} chars of code", file=sys.stderr)
        
        # Execute code
        stdout, stderr, returncode = execute_code(code)
        
        # Output results
        if stdout:
            print(stdout)
        if stderr:
            print(f"[Error] {stderr}", file=sys.stderr)
        
        sys.exit(returncode)
        
    except Exception as e:
        print(f"[Agent Error] {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: orchestration-os-merged/scripts/ruby/agent.rb
#!/usr/bin/env ruby
# Ruby AI Agent - CLI Tool
# Accepts natural language prompts and executes Ruby code

require 'json'
require 'net/http'
require 'uri'

OLLAMA_HOST = ENV.fetch('OLLAMA_HOST', 'http://localhost:11434')
OLLAMA_MODEL = ENV.fetch('OLLAMA_MODEL', 'qwen2.5-coder:3b')
GEMINI_API_KEY = ENV['GEMINI_API_KEY']

$stderr.puts "[Agent] Starting Ruby AI Agent"

def call_ollama(prompt)
  uri = URI("#{OLLAMA_HOST}/api/chat")
  req = Net::HTTP::Post.new(uri)
  req['Content-Type'] = 'application/json'
  req.body = {
    model: OLLAMA_MODEL,
    messages: [{ role: 'user', content: prompt }],
    stream: false
  }.to_json

  client = Net::HTTP.new(uri.hostname, uri.port)
  client.open_timeout = 120
  client.read_timeout = 120
  response = client.request(req)
  result = JSON.parse(response.body)
  result.dig('message', 'content')
rescue => e
  $stderr.puts "[Ollama Error] #{e.message}"
  nil
end

def call_gemini(prompt)
  return nil if GEMINI_API_KEY.nil? || GEMINI_API_KEY.empty?

  uri = URI("https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=#{GEMINI_API_KEY}")
  req = Net::HTTP::Post.new(uri)
  req['Content-Type'] = 'application/json'
  req.body = {
    contents: [{ parts: [{ text: prompt }] }],
    generationConfig: { temperature: 0.7, maxOutputTokens: 8192 }
  }.to_json

  response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |h| h.request(req) }
  result = JSON.parse(response.body)
  result.dig('candidates', 0, 'content', 'parts', 0, 'text')
rescue => e
  $stderr.puts "[Gemini Error] #{e.message}"
  nil
end

def generate_code(prompt)
  full_prompt = <<~PROMPT
    You are an expert Ruby programmer. Generate ONLY Ruby code for this request:

    #{prompt}

    RULES:
    - Output ONLY Ruby code, nothing else
    - NO explanations, NO comments describing what the code does
    - NO markdown formatting, NO backticks, NO code block markers
    - Code must be valid Ruby syntax
    - Code must be complete and runnable
    - Use Ruby's built-in methods and standard library
    - For output use: puts, print

    Output just the raw Ruby code now:
  PROMPT

  # Try Ollama first
  result = call_ollama(full_prompt)
  return clean_code(result) if result && !result.empty?

  # Fallback to Gemini
  result = call_gemini(full_prompt)
  return clean_code(result) if result && !result.empty?

  raise 'No LLM available - configure OLLAMA_HOST or GEMINI_API_KEY'
end

def clean_code(raw)
  return '' if raw.nil? || raw.empty?

  code = raw.dup
  code.gsub!(/```ruby/, '')
  code.gsub!(/```ruby/, '')
  code.gsub!(/```\n?/, '')
  code.gsub!(/`/, '')
  code.strip
end

def execute_code(code)
  require 'tempfile'
  
  tmp = Tempfile.new(['agent_', '.rb'])
  begin
    tmp.write(code)
    tmp.close
    
    output = `ruby #{tmp.path} 2>&1`
    exit_code = $?.exitstatus
    [output, '', exit_code]
  ensure
    tmp.unlink
  end
end

def show_help
  puts <<~HELP
    Ruby AI Agent - Generate and execute Ruby code using AI

    Usage: agent.rb <prompt>
           agent.rb --test

    Examples:
      agent.rb "Print hello world 10 times"
      agent.rb "Read a file and count its lines"
      agent.rb --test

    Environment Variables:
      OLLAMA_HOST      Ollama server URL (default: http://localhost:11434)
      OLLAMA_MODEL     Model to use (default: qwen2.5-coder:3b)
      GEMINI_API_KEY   Google Gemini API key (optional)
  HELP
end

def test_llm
  $stderr.puts '[Agent] Testing LLM connectivity...'

  if call_ollama('Say "Hello" in exactly one word')
    $stderr.puts '[Agent] Ollama connection: OK'
  else
    $stderr.puts '[Error] Ollama connection: FAILED'
  end

  if GEMINI_API_KEY && !GEMINI_API_KEY.empty?
    if call_gemini('Say "Hello" in exactly one word')
      $stderr.puts '[Agent] Gemini connection: OK'
    else
      $stderr.puts '[Error] Gemini connection: FAILED'
    end
  else
    $stderr.puts '[Agent] Gemini: Not configured'
  end
end

if __FILE__ == $PROGRAM_NAME
  if ARGV.empty?
    show_help
    exit 1
  end

  if ARGV[0] == '--test'
    test_llm
    exit 0
  end

  prompt = ARGV.join(' ')
  $stderr.puts "[Agent] Generating code for: #{prompt}"

  begin
    code = generate_code(prompt)
    $stderr.puts "[Agent] Generated #{code.length} chars of code"

    stdout, _stderr, exit_code = execute_code(code)

    print stdout
    exit exit_code
  rescue => e
    $stderr.puts "[Agent Error] #{e.message}"
    exit 1
  end
end
